### PR TITLE
Fix line endings handling for downloaded web pages #10268

### PR DESCRIPTION
**Pull request by lkho**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### What is the purpose of your *pull request*?
- [x] Bug fix
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Suggested fix for #10268

### CODE REVIEW

**Comment by yan12125** [comment]
Line endings should never be changed as ```_webpage_read_content``` can be used to read arbitrary binary data:
```
$ python
Python 3.5.2 (default, Jun 28 2016, 08:46:01) 
[GCC 6.1.1 20160602] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import io
>>> webpage_bytes = b'\r\n'
>>> encoding = 'utf-8'
>>> with io.TextIOWrapper(io.BytesIO(webpage_bytes), encoding, 'replace') as intxt:
...     content = intxt.read()
... 
>>> content
'\n'
```

**Comment by lkho** [comment]
How about replacing '\r\n', '\r' with '\n' before writing the subtitle file, such as in https://github.com/rg3/youtube-dl/blob/master/youtube_dl/YoutubeDL.py#L1594 ?

**Comment by yan12125** [comment]
I guess the problem is this line: https://github.com/rg3/youtube-dl/blob/7aa589a/youtube_dl/YoutubeDL.py#L1606. Could you try it to:
```
with io.open(encodeFilename(sub_filename), 'w', encoding='utf-8', newline='') as subfile:
```
And check whether subtitles work or not?

**Comment by lkho** [comment]
It works, the line endings are written untranslated. But will it cause OS compatibility issues?

**Comment by yan12125** [comment]
> will it cause OS compatibility issues?

Did you mean files with UNIX line endings are written as is, so some Windows applications can't handle it?

**Comment by lkho** [comment]
yes, or vice versa

**Comment by yan12125** [comment]
Good question. Take subtitles for example. Assume there's a subtitle file on the Internet with UNIX line endings. I download it with my browser. I would expect it can be handled by media players on Windows without line endings conversion. I guess keeping all contents un-changed is the safest way.

**Comment by lkho** [comment]
>  I would expect it can be handled by media players on Windows without line endings conversion.
good point, good players should handle that well.

**Comment by yan12125** [comment]
There are lots of ```io.open``` usages in this project. Would you like to replace all occurrences?

File: youtube_dl/extractor/common.py [hunk 1]
```diff
@@ -11,6 +11,7 @@
 import sys
 import time
 import math
+import io
```
Comment (yan12125) [inline youtube_dl/extractor/common.py]: Remove this line.

Comment (lkho) [inline youtube_dl/extractor/common.py]: fixed

**Comment by yan12125** [comment]
Could you squash all commits into a single one?

**Comment by lkho** [comment]
like this?

**Comment by yan12125** [comment]
Merged. Thanks!

**Pull request merged**